### PR TITLE
Add initial jq-based templating engine

### LINKED ISSUE #341: Schemantic error in verification no system-ruby packages installed Dockerfile
This line

https://github.com/docker-library/ruby/blob/e040029c82501556371e9d84a9db607dfb1bba51/Dockerfile-slim.template#L101-L102

won't work as expected. Because of the `!` it will always succeed.

Shellcheck tracks this behavior with

```bash
! dpkg -l | grep -i ruby
^-- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.
```

I can't see a clean one-liner to solve this, I think the correct way is to rather run:

`if dpkg -l | grep -i 'ruby'; then return 1; fi`

also `dpkg -l` can list previously uninstalled packages, to verify a package is actually installed it would rather be `| grep -iP '^ii.*ruby.*'`

### PR DESCRIPTION
**Pull request by tianon**
See also docker-library/php#1052 (and linked PRs).

Fixes #341
Closes #342
Closes #348

~~Hopefully this can help make #348 more straightforward/maintainable :pray: :sweat_smile:~~ :tada:

### CODE REVIEW

**Comment by tianon** [comment]
I cherry-picked/adapted #348 into 60b294c3cab56a97c3c893d234ebe9a507313f5e and I think this is much more maintainable (with the same end result). :sweat_smile:

Once the next release of 2.7 drops, we can just drop that section, and we can axe the whole thing if 2.6 receives the same treatment eventually. :smile:

File: update.sh [hunk 1]
```diff
@@ -3,102 +3,5 @@ set -Eeuo pipefail
...
-
-		# https://github.com/docker-library/ruby/issues/246
-		if [ "$rcVersion" = '2.5' ]; then
-			rubygems='3.0.3'
-			sed -ri \
-				-e 's!%%RUBYGEMS%%!'"$rubygems"'!g' \
-				"$dir/Dockerfile"
-		else
-			sed -ri -e '/RUBYGEMS_VERSION/d' "$dir/Dockerfile"
-		fi
```
Comment (tianon) [inline update.sh:102]: :eyes: :rocket: :sun_with_face:

**Comment by tianon** [comment]
Here's the money shot: (stripping off the second commit so the diff is more obvious)

<details>
<summary>Diff:</summary>

```diff
$ diff -u <(bashbrew cat ruby) <(bashbrew cat <(./generate-stackbrew-library.sh))
--- /dev/fd/63	2021-07-02 16:51:20.941894259 -0700
+++ /dev/fd/62	2021-07-02 16:51:20.941894259 -0700
@@ -3,70 +3,70 @@
 
 Tags: 3.0.1-buster, 3.0-buster, 3-buster, buster, 3.0.1, 3.0, 3, latest
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 3.0/buster
 
 Tags: 3.0.1-slim-buster, 3.0-slim-buster, 3-slim-buster, slim-buster, 3.0.1-slim, 3.0-slim, 3-slim, slim
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
-Directory: 3.0/buster/slim
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
+Directory: 3.0/slim-buster
 
 Tags: 3.0.1-alpine3.13, 3.0-alpine3.13, 3-alpine3.13, alpine3.13, 3.0.1-alpine, 3.0-alpine, 3-alpine, alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 3.0/alpine3.13
 
 Tags: 3.0.1-alpine3.12, 3.0-alpine3.12, 3-alpine3.12, alpine3.12
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 3.0/alpine3.12
 
 Tags: 2.7.3-buster, 2.7-buster, 2-buster, 2.7.3, 2.7, 2
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.7/buster
 
 Tags: 2.7.3-slim-buster, 2.7-slim-buster, 2-slim-buster, 2.7.3-slim, 2.7-slim, 2-slim
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
-Directory: 2.7/buster/slim
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
+Directory: 2.7/slim-buster
 
 Tags: 2.7.3-alpine3.13, 2.7-alpine3.13, 2-alpine3.13, 2.7.3-alpine, 2.7-alpine, 2-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.7/alpine3.13
 
 Tags: 2.7.3-alpine3.12, 2.7-alpine3.12, 2-alpine3.12
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.7/alpine3.12
 
 Tags: 2.6.7-buster, 2.6-buster, 2.6.7, 2.6
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.6/buster
 
 Tags: 2.6.7-slim-buster, 2.6-slim-buster, 2.6.7-slim, 2.6-slim
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
-Directory: 2.6/buster/slim
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
+Directory: 2.6/slim-buster
 
 Tags: 2.6.7-stretch, 2.6-stretch
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.6/stretch
 
 Tags: 2.6.7-slim-stretch, 2.6-slim-stretch
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
-Directory: 2.6/stretch/slim
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
+Directory: 2.6/slim-stretch
 
 Tags: 2.6.7-alpine3.13, 2.6-alpine3.13, 2.6.7-alpine, 2.6-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.6/alpine3.13
 
 Tags: 2.6.7-alpine3.12, 2.6-alpine3.12
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 3904524e5d9e538b33525a602a4bcb7618aff8b6
+GitCommit: e1bd3d07b1b92355bf5358e83732c649c1f93c9f
 Directory: 2.6/alpine3.12
```

</details>

**Review by yosifkit — APPROVED** [review]

**Comment by tianon** [comment]
Someone(tm) should probably file an issue about not having `g++` installed causing Ruby 3.0.2 on Alpine 3.14 to fail to `./configure` (even though it doesn't appear to be used)... :grimacing:

File: 3.0/alpine3.14/Dockerfile [hunk 2]
```diff
@@ -1,7 +1,21 @@
...
+RUN set -eux; \
+	apk add --no-cache \
+		bzip2 \
+		ca-certificates \
+		gmp-dev \
+		libffi-dev \
+		procps \
+		yaml-dev \
+		zlib-dev \
+	;
```
Comment (tianon) [inline 3.0/alpine3.14/Dockerfile:18]: In the context of https://github.com/docker-library/ruby/issues/492, I'm trying to remember why we added all this here, and I cannot for the life of me remember why we added all these permanently. :sob:

File: 3.0/alpine3.14/Dockerfile [hunk 3]
```diff
@@ -95,24 +111,19 @@ RUN set -eux; \
...
-		$runDeps \
-		bzip2 \
-		ca-certificates \
-		libffi-dev \
-		procps \
-		yaml-dev \
-		zlib-dev \
-	; \
+	apk add --no-network --virtual .ruby-rundeps $runDeps; \
 	apk del --no-network .ruby-builddeps; \
```
Comment (tianon) [inline 3.0/alpine3.14/Dockerfile:115]: Aha, this is where they moved from - the `rundeps`.

**Pull request merged**